scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowFunctions.scala

```diff
@@ -263,12 +270,96 @@ class TFExampleSCollectionFunctions[T <: Example](val self: SCollection[T]) {
    *           [[com.spotify.featran.FeatureSpec]]
    * @group output
    */
+  @deprecated("TFRecordSpec will be removed in favor of tf.metadata Schema. Use " +
+    "'saveAsTfExampleFileWithMetadata'", "scio-tensorflow 0.5.6")
   def saveAsTfExampleFile(path: String,
                           fe: FeatureExtractor[SCollection, _],
                           compression: Compression): (Future[Tap[Example]], Future[Tap[String]]) =
     self.saveAsTfExampleFile(path,
                              FeatranTFRecordSpec.fromFeatureSpec(fe.featureNames),
                              compression = compression)
+
+
+  /**
+   * Save this SCollection of `org.tensorflow.example.Example` as a TensorFlow TFRecord file,
+   * along with a protobuf file representing the inferred `org.tensorflow.metadata.v0.Schema`.
+   * @group output
+   */
+  def saveAsTfExampleFileWithMetadata(path: String,
+                                      suffix: String = ".tfrecords",
+                                      compression: Compression = Compression.UNCOMPRESSED,
+                                      numShards: Int = 0)
+                                     (implicit ev: T <:< Example): Future[Tap[Example]] = {
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    buildAndSaveExampleMetadata(path)(ev)
+    val r = self.map(_.toByteArray).saveAsTfRecordFile(path, suffix, compression, numShards)
+    r.map(_.map(Example.parseFrom))
+  }
+
+  /**
+   * Infer a `org.tensorflow.metadata.v0.Schema` from this SCollection of
+   * `org.tensorflow.example.Example`.
+   * @return A singleton `SCollection` containing the schema.
+   */
+  def buildExampleMetadata(implicit ev: T <:< Example): SCollection[Schema] = {
+    val features = examplesToFeatures(self.asInstanceOf[SCollection[Example]])
+    features
+      .groupBy(_ => ())
+      .values.map(features => Schema.newBuilder().addAllFeature(features.asJava).build())
+  }
+
+  /**
+   * Infer a `org.tensorflow.metadata.v0.Schema` from this SCollection of
+   * `org.tensorflow.example.Example` and output the resulting protobuf object.
+   * @return A singleton `SCollection` containing the schema.
+   * @group output
+   */
+  def buildAndSaveExampleMetadata(path: String)(implicit ev: T <:< Example): SCollection[Schema] = {
+    val schema = self.buildExampleMetadata(ev)
+    schema.map { s =>
+      val d = FileSystems.matchNewResource(path, false)
+      val chnnl = Channels.newOutputStream(FileSystems.create(d, MimeTypes.BINARY))
+      try {
+        s.writeTo(chnnl)
+      } finally {
```

Comment:
@ravwojdyla We should definitely handle the failure scenario better here. What do you think the expected behavior should be? Maybe we could add retries or something, but if it fails after that we'd kinda want the job to fail right?